Client/trunk/ParaEngineClient/BlockEngine/BlockRegion.cpp

```diff
@@ -899,172 +905,206 @@ namespace ParaEngine
+						memFile.WriteDWORD(blockLightMapData.size());
+						memFile.write(&(blockLightMapData[0]), blockLightMapData.size());
 					}
 				}
-				
-				if(blockData.size() > 0 || ((dwChunkDataMask & ChunkDataMask_HasCustomData) == 0) )
+
+				if (blockData.size() > 0 || ((dwChunkDataMask & ChunkDataMask_HasCustomData) == 0))
 				{
 					//block count
-					cfile.WriteDWORD(blockData.size());
+					memFile.WriteDWORD(blockData.size());
 					//block id data
-					if(blockData.size() > 0)
+					if (blockData.size() > 0)
 					{
-						std::pair<uint16_t,uint16_t>* pBlockData = &blockData[0];
-						cfile.write(pBlockData, blockData.size() * dataItemSize);
+						std::pair<uint16_t, uint16_t>* pBlockData = &blockData[0];
+						memFile.write(pBlockData, blockData.size() * dataItemSize);
 					}
 				}
 
-				if(blockMetaData.size() > 0)
+				if (blockMetaData.size() > 0)
 				{
 					//block count
-					cfile.WriteDWORD(blockMetaData.size());
+					memFile.WriteDWORD(blockMetaData.size());
 					// block meta data
-					std::pair<uint16_t,uint16_t>* pBlockData = &blockMetaData[0];
-					cfile.write(pBlockData, blockMetaData.size() * dataItemSize);
+					std::pair<uint16_t, uint16_t>* pBlockData = &blockMetaData[0];
+					memFile.write(pBlockData, blockMetaData.size() * dataItemSize);
 				}
 			}
-			cfile.close();
+			auto pStringBuilder = static_cast<StringBuilder*>(memFile.GetHandlePtr());
 
-			GetBlockWorld()->OnSaveBlockRegion(m_regionX, m_regionZ);
-		}
-	}
+			// save compressed data
+			do
+			{			
+				auto uncompressedSize = pStringBuilder->size();
 
-	void BlockRegion::LoadFromFile()
-	{
-		CParaFile* pFile = nullptr;
-		std::string fileName = m_pBlockWorld->GetWorldInfo().GetBlockRegionFileName(m_regionX,m_regionZ,true);
+				// 
+				if (uncompressedSize > 1024)
+				{
+					char* compressedData = new char[uncompressedSize];
+					size_t compressedSize = 0;
 
-		CParaFile gameSaveFile;
-		CParaFile tempFile;
-		SetModified(false);
-		tempFile.OpenAssetFile(fileName.c_str(),true,ParaTerrain::Settings::GetInstance()->GetMediaPath());
-		if(!tempFile.isEof())
-			pFile = &tempFile;
+					z_stream stream;
+					stream.next_in = (Bytef*)pStringBuilder->str();
+					stream.avail_in = (uInt)uncompressedSize;
 
-		if(!pFile)
-		{
-			fileName = m_pBlockWorld->GetWorldInfo().GetBlockRegionFileName(m_regionX, m_regionZ, false);
-			gameSaveFile.OpenAssetFile(fileName.c_str(),true,ParaTerrain::Settings::GetInstance()->GetMediaPath());
-			if(!gameSaveFile.isEof())
-				pFile = &gameSaveFile;
-		}
+					stream.next_out = (Bytef*)compressedData;
+					stream.avail_out = (uInt)uncompressedSize;
 
-		if(pFile == nullptr)
-		{
-			m_nEventAsyncLoadWorldFinished = 2;
-			return;
-		}
+					stream.zalloc = (alloc_func)0;
+					stream.zfree = (free_func)0;
+					stream.opaque = (voidpf)0;
 
-		uint32_t fileTypeId = pFile->ReadDWORD();
-		if(fileTypeId != 0x626c6f63)
-			return;
+					int err = deflateInit(&stream, Z_DEFAULT_COMPRESSION);
```

Comment:
i already wrapped zip/unzip in to a helper class, which you may use directly

https://github.com/LiXizhi/NPLRuntime/wiki/FilesAndIO#compress-and-decompress-with-zlib-and-gzip